### PR TITLE
Refactor: optimize various code strutures and operator implementation

### PR DESCRIPTION
**Pull request by Ziminli**
* refactor: change all `cudaMalloc`, `cudaMemcpy` and `cudaFree` to their async counterparts

* refactor: remove the `cudaGetDeviceProperties` used in elementwise.cu and softmax.cu for performance improvement (currently are replaced with reasonable hard-coded values)

* refactor: combine the device memory allocations and frees into a single one for operators that requires more than 1 set of those

* feat: add profiling for gpt2 cuda training example, which shows the elapsed time and tokens per second for each step

* refactor: optimize the implementation for both the forward and backward CUDA kernels of the cross entropy operator

Screenshot for passing mnist cuda example (only shows a portion at the end):
image

Screenshot for passing gpt2 cuda example: 
image

### CODE REVIEW

**Comment by kilinchange** [comment]
请贴一下 gpt2 和 mnist 端到端训练输出的截图

**Review by kilinchange — APPROVED** [review]

**Review by Chamberlain0w0 — APPROVED** [review]
